File: Algoritmic_Heights/MAJ/MAJ.py
file_path = './MAJ_input.txt'

with open(file_path, 'r') as f:
	a, b = list(map(int, f.readline().strip().split()))
	
	arrays = []
	for i in range(a):
		new_arr = list(map(int, f.readline().strip().split()))
		arrays.append(new_arr)

#realisation of Boyer–Moore string-search algorithm
def find_majority_element(arr):
    candidate = None
    count = 0

    for num in arr:
        if count == 0:
            candidate = num
            count = 1
        elif num == candidate:
            count += 1
        else:
            count -= 1

    count = 0
    for num in arr:
        if num == candidate:
            count += 1

    if count > len(arr) // 2:
        return candidate
    else:
        return -1
       
results = []
for i in arrays:
	results.append(find_majority_element(i))

print(*results)
with open('./MAJ_output.txt', 'w') as f:
	result_string = ' '.join(map(str, results))
	f.write(result_string)
		

File: Algoritmic_Heights/MAJ/MAJ.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <string>

int find_majority_element(const std::vector<int>& arr) {
    int candidate = nullptr;
    int count = 0;

    for (int num : arr) {
        if (count == 0) {
            candidate = num;
            count = 1;
        } else if (num == candidate) {
            count += 1;
        } else {
            count -= 1;
        }
    }

    count = 0;
    for (int num : arr) {
        if (num == candidate) {
            count += 1;
        }
    }

    if (count > arr.size() / 2) {
        return candidate;
    } else {
        return -1;
    }
}

int main() {
    std::ifstream input_file("./MAJ_input.txt");

    int a, b;
    input_file >> a >> b;

    std::vector<std::vector<int>> arrays;
    for (int i = 0; i < a; ++i) {
        std::string line;
        std::getline(input_file, line);
        std::istringstream iss(line);
        std::vector<int> new_arr;
        int num;
        while (iss >> num) {
            new_arr.push_back(num);
        }
        arrays.push_back(new_arr);
    }

    input_file.close();

    std::vector<int> results;
    for (const auto& arr : arrays) {
        results.push_back(find_majority_element(arr));
    }

    for (int result : results) {
        std::cout << result << " ";
    }
    std::cout << std::endl;

    std::ofstream output_file("./MAJ_output.txt");
    for (int result : results) {
        output_file << result << " ";
    }
    output_file.close();

    return 0;
}
